# Explore here

In [1]:
import pandas as pd
import pickle

from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer

from sklearn.naive_bayes import GaussianNB, MultinomialNB, BernoulliNB
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import accuracy_score, classification_report

In [2]:
url = "https://raw.githubusercontent.com/4GeeksAcademy/naive-bayes-project-tutorial/main/playstore_reviews.csv"

df = pd.read_csv(url)

df.head()

,package_name,review,polarity
0,com.facebook.katana,privacy at least put some option appear offli...,0
1,com.facebook.katana,"messenger issues ever since the last update, ...",0
2,com.facebook.katana,profile any time my wife or anybody has more ...,0
3,com.facebook.katana,the new features suck for those of us who don...,0
4,com.facebook.katana,forced reload on uploading pic on replying co...,0


## Paso 2: Estudio de variables y su contenido

In [3]:
df = df.drop(columns=["package_name"])

df["review"] = df["review"].str.strip().str.lower()

df.head()

,review,polarity
0,privacy at least put some option appear offlin...,0
1,"messenger issues ever since the last update, i...",0
2,profile any time my wife or anybody has more t...,0
3,the new features suck for those of us who don'...,0
4,forced reload on uploading pic on replying com...,0


In [4]:
X = df["review"]
y = df["polarity"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [5]:
vec_model = CountVectorizer(stop_words="english")

X_train = vec_model.fit_transform(X_train).toarray()
X_test = vec_model.transform(X_test).toarray()

## Paso 3: Construye un Naive Bayes

In [6]:
multinomial_model = MultinomialNB()
multinomial_model.fit(X_train, y_train)

y_pred_multinomial = multinomial_model.predict(X_test)

accuracy_multinomial = accuracy_score(y_test, y_pred_multinomial)

print("Accuracy (MultinomialNB):", accuracy_multinomial)

Accuracy (MultinomialNB): 0.8156424581005587


In [7]:
gaussian_model = GaussianNB()
gaussian_model.fit(X_train, y_train)

y_pred_gaussian = gaussian_model.predict(X_test)

accuracy_gaussian = accuracy_score(y_test, y_pred_gaussian)

print("Accuracy (GaussianNB):", accuracy_gaussian)

Accuracy (GaussianNB): 0.8044692737430168


In [8]:
bernoulli_model = BernoulliNB()
bernoulli_model.fit(X_train, y_train)

y_pred_bernoulli = bernoulli_model.predict(X_test)

accuracy_bernoulli = accuracy_score(y_test, y_pred_bernoulli)

print("Accuracy (BernoulliNB):", accuracy_bernoulli)

Accuracy (BernoulliNB): 0.770949720670391


MultinomialNB fue el modelo con mejor desempeño, alcanzando una exactitud (accuracy) de aproximadamente 81.56 %, superior a GaussianNB y BernoulliNB. Debido a que los datos fueron transformados mediante CountVectorizer, que genera conteos de palabras, esta implementación resulta la más adecuada para este problema de clasificación de texto.

## Paso 4: Optimiza el modelo anterior

In [9]:
random_forest_model = RandomForestClassifier(random_state=42)

random_forest_model.fit(X_train, y_train)

y_pred_random_forest = random_forest_model.predict(X_test)

accuracy_random_forest = accuracy_score(y_test, y_pred_random_forest)

print("Accuracy (Random Forest):", accuracy_random_forest)

Accuracy (Random Forest): 0.7988826815642458


Random Forest obtuvo una exactitud (accuracy) aproximada de **79.89 %**, inferior al **81.56 %** alcanzado por MultinomialNB. Por lo tanto, no fue posible mejorar el rendimiento del modelo original mediante Random Forest y se mantiene **MultinomialNB** como la mejor alternativa para este conjunto de datos.

## Paso 5: Guarda el modelo

In [10]:
pickle.dump(multinomial_model, open("../models/multinomial_nb_classifier.sav", "wb"))

## Paso 6: Explora otras alternativas

In [11]:
logistic_model = LogisticRegression(random_state=42, max_iter=1000)

logistic_model.fit(X_train, y_train)

y_pred_logistic = logistic_model.predict(X_test)

accuracy_logistic = accuracy_score(y_test, y_pred_logistic)

print("Accuracy (Logistic Regression):", accuracy_logistic)

Accuracy (Logistic Regression): 0.8324022346368715


La Regresión Logística obtuvo una exactitud (accuracy) aproximada de **83.24 %**, superando el rendimiento de MultinomialNB (**81.56 %**). Debido a su mejor desempeño en este conjunto de datos, se presenta como una alternativa más adecuada para este problema de clasificación de texto.